proposal/first-class-calc.md

```diff
@@ -0,0 +1,536 @@
+
+<x><pre>
+**CalcExpression** ::= `calc(`¹ CalcArgument ')'
+**ClampExpression** ::= `clamp(`¹ CalcArgument ',' CalcArgument ',' CalcArgument ')'
+**CalcArgument**²  ::= InterpolatedDeclarationValue | CalcSum
+**CalcSum**     ::= CalcProduct (('+' | '-')³ CalcProduct)?
+**CalcProduct** ::= CalcValue (('*' | '/') CalcValue)?
+**CalcValue**   ::= '(' CalcArgument ')
+&#32;             | CalcExpression
+&#32;             | ClampExpression
+&#32;             | CssMinMax
+&#32;             | FunctionExpression⁴
+&#32;             | Number
+&#32;             | Variable
+</pre></x>
+
+1: The strings `calc(` and `clamp(` are matched case-insensitively.
+
+2: A `CalcArgument` is only parsed as an `InterpolatedDeclarationValue` if it
+includes interpolation, unless that interpolation is within a region bounded by
+parentheses (a `FunctionExpression` counts as parentheses).
+
+3: Whitespace is required around these `"+"` and `"-"` tokens.
+
+4: This `FunctionExpression` cannot begin with `min(`, `max(`, or `clamp(`,
+case-insensitively.
+
+[`<function-token>`]: https://drafts.csswg.org/css-syntax-3/#ref-for-typedef-function-token%E2%91%A0
+
+> The `CalcArgument` production provides backwards-compatibility with the
+> historical use of interpolation to inject SassScript values into `calc()`
+> expressions. Because interpolation could inject any part of a `calc()`
+> expression regardless of syntax, for full compatibility it's necessary to
+> parse it very expansively.
+>
+> Note that the interpolation in the definition of `CalcValue` is only reachable
+> from a `CssMinMax` production, *not* from `CalcExpression`. This is
+> intentional, for backwards-compatibility with the existing `CssMinMax` syntax.x
+
+### `CssMinMax`
+
+This proposal replaces the reference to `CalcValue` in the definition of
+[`CssMinMax`] with `CalcArgument`.
+
+[`CssMinMax`]: ../spec/syntax.md#minmaxexpression
+
+> Note that this increases the number of cases where a `MinMaxExpression` will
+> be parsed as a `CssMinMax` rather than a `FunctionExpression` (for example,
+> `min($foo, $bar)` is now a valid `CssMinMax` where it wasn't before).
+> Fortunately, this is backwards-compatible, since all such `MinMaxExpression`s
+> that were already valid will be simplified down into the same number they
+> returned before.
+
+## Types
+
+This proposal introduces a new value type known as a "calc", with the following
+structure:
+
+```ts
+interface Calc {
+  name: string;
+  arguments: CalcValue[];
+}
+
+type CalcValue =
+  | Number
+  | UnquotedString
+  | CalcOperation
+  | Calc;
+
+interface CalcOperation {
+  operator: '+' | '-' | '*' | '/';
+  left: CalcValue;
+  right: CalcValue;
+}
+```
+
+Unless otherwise specified, when this specification creates a calc, its name is
+"calc".
+
+### Operations
+
+A calc follows the default behavior of all SassScript operations, except that it
+throws an error if used as an operand of a unary or binary `+` or `-` operation.
```

Comment:
what about the `*` operator ?

Reply:
The default behavior for `*`, `/` and `%` is to throw an error. `+` and `-` are special in that they default to converting their operands to strings.